app/controllers/incomeSources/add/BusinessAddedObligationsController.scala

```diff
@@ -0,0 +1,167 @@
+                                                   val obligationsView: BusinessAddedObligations)
+                                                  (implicit val appConfig: FrontendAppConfig,
+                                                   implicit val itvcErrorHandlerAgent: AgentItvcErrorHandler,
+                                                   implicit override val mcc: MessagesControllerComponents,
+                                                   val ec: ExecutionContext,
+                                                   dateService: DateServiceInterface)
+  extends ClientConfirmedController with I18nSupport with FeatureSwitching{
+
+  lazy val backUrl: String = controllers.incomeSources.add.routes.AddBusinessReportingMethod.show().url
+  private lazy val agentBackUrl = controllers.incomeSources.add.routes.AddBusinessReportingMethod.showAgent().url
+
+  def show(id: Option[String]): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(isAgent = false, id)
+  }
+
+  def showAgent(id: Option[String]): Action[AnyContent] =
+    Authenticated.async {
+      implicit request =>
+        implicit user =>
+          getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+            implicit mtdItUser =>
+              handleRequest(isAgent = true, id)
+          }
+    }
+
+  def handleRequest(isAgent: Boolean, id: Option[String])(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
+    if (isDisabled(IncomeSources)) {
+      Future.successful(Redirect(controllers.routes.HomeController.show()))
+    } else {
+      id match {
+        case None => Logger("application").error(
+          s"[BusinessAddedObligationsController][handleRequest] - Error: No id supplied by reporting method page")
+          Future(itvcErrorHandler.showInternalServerError())
+        case Some(value) =>
+          val addedBusiness: BusinessDetailsModel = user.incomeSources.businesses.find(x => x.incomeSourceId.contains(value)).get
+
+          if (addedBusiness.tradingName.isEmpty) {
+            Logger("application").error(
+              s"[BusinessAddedObligationsController][handleRequest] - No business name for business with provided id")
+            itvcErrorHandler.showInternalServerError()
+          }
+          if (addedBusiness.tradingStartDate.isEmpty) {
+            Logger("application").error(s"[BusinessAddedObligationsController][handleRequest] - No business start date for business with provided id")
+            Future(itvcErrorHandler.showInternalServerError())
+          }
+
+          val businessName: String = addedBusiness.tradingName.get
+
+          val dates: Seq[DatesModel] = getObligationDates(id.get)
+
+          val quarterlyDates: Seq[DatesModel] = dates.filter(x => x.periodKey.contains("00")).sortBy(_.inboundCorrespondenceFrom)
```

Comment:
its worth to consider refactor this code in to service or model

Reply:
I think in terms of readability this is fine, there's nothing that requires it to be separated, and it's calling a service so keeping it in the connector is fine. I only took it out of the main handleRequest method for readability, it's just a chunk of code, but I think it belongs in the connector